workflow.go

```diff
@@ -0,0 +1,257 @@
+				logger.Info("Billing cycle is up! Time to charge...")
+				timerError = f.Get(ctx, nil)
+			}).
+			// Signal Handler for updating billing address or other customer info
+			AddReceive(workflow.GetSignalChannel(ctx, "update"), func(c workflow.ReceiveChannel, _ bool) {
+				var updateInfo UpdateSignal
+				c.Receive(ctx, &updateInfo)
+				logger.Info("Received update signal", "data", updateInfo)
+				s = &updateInfo.Subscription
+			}).
+			// Signal handler for canceling the subscription
+			AddReceive(workflow.GetSignalChannel(ctx, "cancel"), func(c workflow.ReceiveChannel, _ bool) {
+				var cancel UpdateSignal
+				c.Receive(ctx, &cancel)
+				logger.Info("Received cancel signal", "data", cancel)
+
+				cancelSub = cancel.CancelSubscription
+			})
+
+		err = workflow.Await(ctx, func() bool {
+			logger.Info("Current history length", "History Length", info.GetCurrentHistoryLength())
+			return info.GetCurrentHistoryLength() > EventHistoryThreshold || selector.HasPending()
+		})
+		if err != nil {
+			logger.Error("Awaiting history or selector failed.", "Error", err)
+			return err
+		}
+
+		if info.GetCurrentHistoryLength() > EventHistoryThreshold {
+			canError := workflow.NewContinueAsNewError(ctx, SubscriptionWorkflow, *s)
+			return canError
+		}
+
+		selector.Select(ctx)
+
+		if timerError != nil {
+			logger.Error("Billing timer failed.", "Error", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *SubscriptionInfo) drainSignals(ctx workflow.Context) (*SubscriptionInfo, error) {
+	// If there are *any* cancel signals,
+	signalChannel := workflow.GetSignalChannel(ctx, "cancel")
+	var cancel UpdateSignal
+	ok := signalChannel.ReceiveAsync(&cancel)
+	if ok && cancel.CancelSubscription {
+		workflow.GetLogger(ctx).Info("Received cancel signal.")
+		return &SubscriptionInfo{}, errors.New("cancel")
+	}
+
+	for {
+		var update UpdateSignal
+		signalChannel = workflow.GetSignalChannel(ctx, "update")
+		ok := signalChannel.ReceiveAsync(&update)
+		if !ok {
+			break
+		}
+		workflow.GetLogger(ctx).Info("Received update signal.", "Subscription Info", update.Subscription)
+		// overwriting, but keep whatever the last one was.
+		s = &update.Subscription
+	}
+
+	return s, nil
+}
+
+func (s *SubscriptionInfo) cancelSubscription(ctx workflow.Context) (err error) {
+	var a *Activities
+	err = workflow.ExecuteActivity(ctx, a.SendCancelationEmail, s).Get(ctx, nil)
+	if err != nil {
+		workflow.GetLogger(ctx).Error("SendCancelationEmail Activity failed.", "Error", err)
+		return err
+	}
+	return nil
+}
+
+func (s *SubscriptionInfo) maybeContinueAsNew(ctx workflow.Context, canError error) error {
+	workflow.GetLogger(ctx).Info("Received Continue-As-New, draining and then possibly returning as expected.")
+	s, err := s.drainSignals(ctx)
+
+	// if draining found a cancel signal, don't actually continue-as-new, but finish out canceling
+	if err != nil && err.Error() == "cancel" {
```

Comment:
For more idiomatic Go, if you're gonna use errors for control flow, create a `var ErrCancelled = errors.New("cancel")` and check equality with that